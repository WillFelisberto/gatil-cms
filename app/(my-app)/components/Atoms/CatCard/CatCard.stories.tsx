import type { Meta, StoryObj } from '@storybook/react';

import { Cat } from '@/payload-types';

import { CatCard } from './CatCard';

const meta: Meta<typeof CatCard> = {
  title: 'Atoms/CatCard',
  component: CatCard,
  tags: ['autodocs']
};

export default meta;
type Story = StoryObj<typeof CatCard>;

const whatsappNumber = '(11) 91234-5678';

export const Default: Story = {
  args: {
    whatsappNumber,
    cat: {
      id: 'gato-1',
      nome: 'Luna',
      idade: '3 meses',
      descricao: 'Muito carinhosa, adora brincar com bolinhas de papel.',
      foto: {
        url: '/gato.jpg',
        alt: 'Gata Luna'
      },
      sexo: 'F',
      vacinas: [{ nome: 'V4' }],
      vermifugacoes: [{ nome: 'Vermivet' }],
      castrado: true,
      doencas: 'Nenhuma conhecida',
      observacoesSaude: 'Completamente saudável após resgate.'
    } as Cat
  }
};

export const SemVacinaOuVermifugo: Story = {
  args: {
    whatsappNumber,
    cat: {
      id: 'gato-2',
      nome: 'Biscoito',
      idade: '2 anos',
      sexo: 'M',
      descricao: 'Observador e calmo. Adora dormir no sol.',
      foto: {
        url: '/gato.jpg',
        alt: 'Gato Biscoito'
      },
      vermifugacoes: null,
      castrado: true,
      doencas: null,
      observacoesSaude: null,
      galeria: null,
      vacinas: null,
      dataEntrada: '2025-04-02T12:00:00.000Z',
      adotado: false,
      show: true,
      disponivelParaApadrinhamento: true
    } as Cat
  }
};

export const ComDoencas: Story = {
  args: {
    whatsappNumber,
    cat: {
      id: 'gato-3',
      nome: 'Tigrinho',
      sexo: 'M',
      idade: '1 ano e meio',
      descricao: 'Agitado, curioso e precisa de atenção veterinária contínua.',
      foto: {
        url: '/gato.jpg',
        alt: 'Gato Tigrinho'
      },
      vacinas: null,
      vermifugacoes: null,
      castrado: false,
      doencas: 'Leucemia felina',
      observacoesSaude: 'Toma medicação diariamente.',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    } as Cat
  }
};

export const SemFoto: Story = {
  args: {
    whatsappNumber,
    cat: {
      id: 'gato-4',
      nome: 'Noir',
      idade: '6 meses',
      sexo: 'M',
      descricao: 'Ainda sem foto cadastrada, mas muito fofo!',
      vacinas: [],
      vermifugacoes: [],
      castrado: false,
      doencas: null,
      observacoesSaude: null,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      foto: null
    } as Cat
  }
};

export const ApenasComNome: Story = {
  args: {
    whatsappNumber,
    cat: {
      id: 'gato-5',
      nome: 'Fantasma',
      sexo: 'F',
      foto: null
    } as Cat
  }
};
