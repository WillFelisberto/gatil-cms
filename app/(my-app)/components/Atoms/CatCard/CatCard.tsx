'use client';

import Image from 'next/image';
import Link from 'next/link';
import { FaMars, FaVenus } from 'react-icons/fa';

type Media = {
  url: string;
  alt?: string;
};

export type Cat = {
  id: string;
  nome: string;
  idade?: string | null;
  descricao?: string | null;
  foto?: string | Media | null;
  vacinas?:
    | { nome: string; data?: string | null; observacoes?: string | null; id?: string | null }[]
    | null;
  vermifugacoes?:
    | {
        nome?: string | null;
        data?: string | null;
        observacoes?: string | null;
        id?: string | null;
      }[]
    | null;
  castrado?: boolean | null;
  doencas?: string | null;
  observacoesSaude?: string | null;
  sexo: 'M' | 'F';
};

type Props = {
  cat: Cat;
};

export const CatCard = ({ cat }: Props) => {
  const {
    nome,
    idade,
    descricao,
    foto,
    vacinas,
    vermifugacoes,
    castrado,
    doencas,
    observacoesSaude,
    sexo
  } = cat;

  const imageUrl = typeof foto === 'string' ? foto : foto?.url;
  const altText = typeof foto === 'object' && foto?.alt ? foto.alt : `Foto de ${nome}`;

  const renderCastrado = sexo === 'F' ? '✂️ Castrada' : '✂️ Castrado';
  const renderVacina = sexo === 'F' ? '💉 Vacinada' : '💉 Vacinado';
  const renderVermifugo = sexo === 'F' ? '🪱 Vermifugada' : '🪱 Vermifugado';

  const whatsappMessage = encodeURIComponent(
    `Olá! Tenho interesse em adotar ${sexo === 'F' ? 'a' : 'o'} ${nome}. Poderia me passar mais informações?`
  );
  const whatsappLink = `https://wa.me/5599999999999?text=${whatsappMessage}`; //  TODO:   <- Substituir com o número da ONG

  return (
    <article role="article" className="flex flex-col h-full" data-testid={`cat-card-${cat.id}`}>
      <Link
        href={whatsappLink}
        target="_blank"
        rel="noopener noreferrer"
        aria-label={`Saber mais sobre ${nome}`}
        className="bg-stone-50 rounded-2xl shadow-md overflow-hidden flex flex-col w-full max-w-sm border border-gray-200 hover:shadow-lg transition-all duration-200 group relative"
      >
        <span
          className="absolute top-2 left-2 bg-green-100 text-green-900 text-xs font-semibold px-2 py-1 rounded-full shadow-sm z-10 group-hover:scale-105 transition"
          data-testid="contact-badge"
        >
          Clique aqui para saber mais 🐾
        </span>

        <div className="relative h-60 w-full">
          <Image
            src={imageUrl || '/no-image.jpg'}
            alt={altText}
            fill
            sizes="(max-width: 768px) 100vw, 384px"
            className="object-cover"
            priority
            data-testid="cat-image"
          />
        </div>

        <div className="p-4 flex flex-col gap-2">
          <h2
            className="text-xl font-bold text-blue-950 flex items-center gap-2"
            data-testid="cat-name"
          >
            {sexo === 'M' ? (
              <FaMars className="text-blue-700" title="Macho" aria-label="Macho" />
            ) : (
              <FaVenus className="text-pink-600" title="Fêmea" aria-label="Fêmea" />
            )}
            {nome}
          </h2>
          {descricao && (
            <p
              className="text-sm font-medium font-poppins text-gray-600"
              data-testid="cat-description"
            >
              {descricao}
            </p>
          )}

          <div
            className="mt-2 space-y-1 text-sm font-medium text-gray-600"
            data-testid="cat-details"
          >
            {idade && <p className="text-sm text-gray-700">🎂 {idade}</p>}
            {vacinas && vacinas.length > 0 && <p>{renderVacina}</p>}
            {castrado && <p>{renderCastrado}</p>}
            {vermifugacoes && vermifugacoes.length > 0 && <p>{renderVermifugo}</p>}
            {doencas && <p>⚠️ Doenças: {doencas}</p>}
            {observacoesSaude && <p>🩺 {observacoesSaude}</p>}
          </div>
        </div>
      </Link>
    </article>
  );
};
