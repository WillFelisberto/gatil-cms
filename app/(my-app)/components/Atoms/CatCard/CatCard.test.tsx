import { fireEvent, render, screen } from '@/tests/test-utils';

import { CatCard } from './CatCard';

describe('<CatCard />', () => {
  const baseCat = {
    id: '1',
    nome: 'Gato Teste',
    sexo: 'M' as const,
    createdAt: '2025-04-02T21:24:52.909Z',
    updatedAt: '2025-04-03T20:30:41.034Z',
    idade: '1 ano',
    descricao: 'maluca doidinha',
    foto: {
      createdAt: '2025-04-02T20:48:51.116Z',
      updatedAt: '2025-04-02T20:48:51.116Z',
      alt: 'Will',
      filename: 'gato3.png',
      mimeType: 'image/png',
      filesize: 1008532,
      width: 800,
      height: 800,
      focalX: 50,
      focalY: 50,
      sizes: {
        thumbnail: {
          width: 400,
          height: 300,
          mimeType: 'image/png',
          filesize: 324915,
          filename: 'gato3-400x300.png',
          url: '/api/media/file/gato3-400x300.png'
        },

        card: {
          width: 768,
          height: 1024,
          mimeType: 'image/png',
          filesize: 1794522,
          filename: 'gato3-768x1024.png',
          url: '/api/media/file/gato3-768x1024.png'
        },

        tablet: {
          width: null,
          height: null,
          mimeType: null,
          filesize: null,
          filename: null,
          url: null
        }
      },
      id: '67eda2b3b852bb6b7ae49216',
      url: '/api/media/file/gato3.png',
      thumbnailURL: '/api/media/file/gato3-400x300.png'
    },
    galeria: [],
    vacinas: [],
    vermifugacoes: [],
    castrado: true,
    dataEntrada: '2025-04-02T12:00:00.000Z',
    adotado: false,
    show: true,
    disponivelParaApadrinhamento: true
  };

  const whatsappNumber = '(11) 99999-9999';

  it('renders correctly with required props', () => {
    const { container } = render(<CatCard cat={baseCat} whatsappNumber={whatsappNumber} />);

    expect(screen.getByTestId('cat-name')).toHaveTextContent('Gato Teste');
    expect(screen.getByTestId('contact-badge')).toBeInTheDocument();
    expect(screen.getByTestId('cat-image')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('renders correctly in "adotar" mode with WhatsApp link', () => {
    render(<CatCard cat={baseCat} whatsappNumber={whatsappNumber} mode="adotar" />);

    const expectedMessage = encodeURIComponent(
      `Ol√°! Tenho interesse em adotar o Gato Teste. Poderia me passar mais informa√ß√µes?`
    );

    expect(screen.getByTestId('cat-name')).toHaveTextContent('Gato Teste');
    expect(screen.getByRole('link')).toHaveAttribute(
      'href',
      expect.stringContaining(expectedMessage)
    );
    expect(screen.getByTestId('contact-badge')).toHaveTextContent('Clique aqui para saber mais üêæ');
  });

  it('renders correctly in "apadrinhar" mode and triggers onClick', () => {
    const handleClick = jest.fn();

    render(
      <CatCard
        cat={baseCat}
        whatsappNumber={whatsappNumber}
        mode="apadrinhar"
        onClick={handleClick}
      />
    );

    const card = screen.getByRole('button', { name: /Apadrinhar Gato Teste/i });
    expect(card).toBeInTheDocument();
    expect(screen.getByTestId('contact-badge')).toHaveTextContent('Clique aqui para apadrinhar üêæ');

    fireEvent.click(card);
    expect(handleClick).toHaveBeenCalled();
  });

  it('displays male icon for male cats', () => {
    render(<CatCard cat={baseCat} whatsappNumber={whatsappNumber} />);
    expect(screen.getByTitle('Macho')).toBeInTheDocument();
  });

  it('displays female icon for female cats', () => {
    const femaleCat = { ...baseCat, sexo: 'F' as const };
    render(<CatCard cat={femaleCat} whatsappNumber={whatsappNumber} />);
    expect(screen.getByTitle('F√™mea')).toBeInTheDocument();
  });

  it('displays optional fields when provided', () => {
    const completeCat = {
      ...baseCat,
      idade: '2 anos',
      descricao: 'Gato muito brincalh√£o',
      foto: {
        id: 'test-id',
        createdAt: '2025-04-02T20:48:51.116Z',
        updatedAt: '2025-04-02T20:48:51.116Z',
        url: '/test.jpg',
        alt: 'Foto teste',
        mimeType: 'image/jpeg',
        filesize: 12345,
        width: 800,
        height: 800,
        focalX: 50,
        focalY: 50,
        sizes: {
          thumbnail: {
            width: 400,
            height: 300,
            mimeType: 'image/jpeg',
            filesize: 1234,
            filename: 'test-thumbnail.jpg',
            url: '/test-thumbnail.jpg'
          },
          card: undefined,
          tablet: undefined
        }
      },
      vacinas: [{ nome: 'Raiva' }],
      castrado: true,
      vermifugacoes: [{}],
      doencas: 'Nenhuma',
      observacoesSaude: 'Saud√°vel'
    };

    render(<CatCard cat={completeCat} whatsappNumber={whatsappNumber} />);

    expect(screen.getByText('üéÇ 2 anos')).toBeInTheDocument();
    expect(screen.getByText('Gato muito brincalh√£o')).toBeInTheDocument();
    expect(screen.getByText('üíâ Vacinado')).toBeInTheDocument();
    expect(screen.getByText('‚úÇÔ∏è Castrado')).toBeInTheDocument();
    expect(screen.getByText('ü™± Vermifugado')).toBeInTheDocument();
    expect(screen.getByText('‚ö†Ô∏è Doen√ßas: Nenhuma')).toBeInTheDocument();
    expect(screen.getByText('üìù Saud√°vel')).toBeInTheDocument();
  });

  it('handles different image types correctly', () => {
    const stringPhotoCat = { ...baseCat, foto: '/string-photo.jpg' };
    const objectPhotoCat = {
      ...baseCat,
      foto: {
        id: 'test-id',
        createdAt: '2025-04-02T20:48:51.116Z',
        updatedAt: '2025-04-02T20:48:51.116Z',
        url: '/object-photo.jpg',
        alt: 'Alt texto',
        mimeType: 'image/jpeg',
        filesize: 12345,
        width: 800,
        height: 800,
        focalX: 50,
        focalY: 50,
        sizes: {
          thumbnail: undefined,
          card: undefined,
          tablet: undefined
        }
      }
    };

    const { rerender } = render(<CatCard cat={stringPhotoCat} whatsappNumber={whatsappNumber} />);
    expect(screen.getByTestId('cat-image').getAttribute('src')).toContain('string-photo.jpg');

    rerender(<CatCard cat={objectPhotoCat} whatsappNumber={whatsappNumber} />);
    expect(screen.getByTestId('cat-image').getAttribute('src')).toContain('object-photo.jpg');
    expect(screen.getByTestId('cat-image')).toHaveAttribute('alt', 'Alt texto');
  });

  it('uses fallback image when no photo provided', () => {
    render(<CatCard cat={baseCat} whatsappNumber={whatsappNumber} />);
    expect(screen.getByTestId('cat-image').getAttribute('src')).toContain('no-image.jpg');
    expect(screen.getByTestId('cat-image')).toHaveAttribute('alt', 'Foto de Gato Teste');
  });

  it('generates correct WhatsApp link', () => {
    render(<CatCard cat={baseCat} whatsappNumber={whatsappNumber} />);
    const expectedMessage = encodeURIComponent(
      `Ol√°! Tenho interesse em adotar o Gato Teste. Poderia me passar mais informa√ß√µes?`
    );
    expect(screen.getByRole('link')).toHaveAttribute(
      'href',
      expect.stringContaining(expectedMessage)
    );
    expect(screen.getByRole('link')).toHaveAttribute(
      'href',
      expect.stringContaining('https://wa.me/5511999999999')
    );
  });

  it('displays gender-specific texts correctly', () => {
    const femaleCat = {
      ...baseCat,
      sexo: 'F' as const,
      castrado: true,
      vacinas: [{ nome: 'Raiva' }]
    };

    const { rerender } = render(<CatCard cat={femaleCat} whatsappNumber={whatsappNumber} />);
    expect(screen.getByText('‚úÇÔ∏è Castrada')).toBeInTheDocument();
    expect(screen.getByText('üíâ Vacinada')).toBeInTheDocument();

    rerender(
      <CatCard cat={{ ...femaleCat, vermifugacoes: [{}] }} whatsappNumber={whatsappNumber} />
    );
    expect(screen.getByText('ü™± Vermifugada')).toBeInTheDocument();
  });

  it('does not render optional fields when not provided', () => {
    render(<CatCard cat={baseCat} whatsappNumber={whatsappNumber} />);

    expect(screen.queryByTestId('cat-description')).not.toBeInTheDocument();
    expect(screen.queryByText('üéÇ')).not.toBeInTheDocument();
    expect(screen.queryByText('üíâ Vacinado')).not.toBeInTheDocument();
    expect(screen.queryByText('‚úÇÔ∏è Castrado')).not.toBeInTheDocument();
    expect(screen.queryByText('ü™± Vermifugado')).not.toBeInTheDocument();
    expect(screen.queryByText('‚ö†Ô∏è Doen√ßas:')).not.toBeInTheDocument();
    expect(screen.queryByText('üìù')).not.toBeInTheDocument();
  });

  it('handles empty vaccine and deworming arrays', () => {
    const catWithEmptyArrays = {
      ...baseCat,
      vacinas: [],
      vermifugacoes: []
    };

    render(<CatCard cat={catWithEmptyArrays} whatsappNumber={whatsappNumber} />);

    expect(screen.queryByText('üíâ Vacinado')).not.toBeInTheDocument();
    expect(screen.queryByText('ü™± Vermifugado')).not.toBeInTheDocument();
  });
});
