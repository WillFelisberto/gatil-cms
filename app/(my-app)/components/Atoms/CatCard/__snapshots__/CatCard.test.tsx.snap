// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatCard /> renders correctly with required props 1`] = `
<div>
  <article
    class="h-full w-full max-w-96"
    data-testid="cat-card-1"
    role="article"
  >
    <a
      aria-label="Saber mais sobre Gato Teste"
      class="bg-stone-50 h-full flex flex-col justify-between rounded-2xl shadow-md overflow-hidden w-full border border-gray-200 hover:shadow-lg transition-all duration-200 group relative"
      href="https://wa.me/5599999999999?text=Ol%C3%A1!%20Tenho%20interesse%20em%20adotar%20o%20Gato%20Teste.%20Poderia%20me%20passar%20mais%20informa%C3%A7%C3%B5es%3F"
      rel="noopener noreferrer"
      target="_blank"
    >
      <span
        class="absolute top-2 left-2 bg-green-100 text-green-900 text-xs font-semibold px-2 py-1 rounded-full shadow-sm z-10 group-hover:scale-105 transition"
        data-testid="contact-badge"
      >
        Clique aqui para saber mais 🐾
      </span>
      <div
        class="relative h-60 w-full shrink-0"
      >
        <img
          alt="Foto de Gato Teste"
          class="object-cover"
          data-nimg="fill"
          data-testid="cat-image"
          decoding="async"
          sizes="(max-width: 768px) 100vw, 384px"
          src="/_next/image?url=%2Fno-image.jpg&w=3840&q=75"
          srcset="/_next/image?url=%2Fno-image.jpg&w=640&q=75 640w, /_next/image?url=%2Fno-image.jpg&w=750&q=75 750w, /_next/image?url=%2Fno-image.jpg&w=828&q=75 828w, /_next/image?url=%2Fno-image.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fno-image.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fno-image.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fno-image.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fno-image.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="p-4 h-full flex flex-col gap-2"
      >
        <h2
          class="text-xl font-bold  flex items-center gap-2 text-blue-700"
          data-testid="cat-name"
        >
          <svg
            aria-label="Macho"
            class="text-blue-700"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 384 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Macho
            </title>
            <path
              d="M372 64h-79c-10.7 0-16 12.9-8.5 20.5l16.9 16.9-80.7 80.7c-22.2-14-48.5-22.1-76.7-22.1C64.5 160 0 224.5 0 304s64.5 144 144 144 144-64.5 144-144c0-28.2-8.1-54.5-22.1-76.7l80.7-80.7 16.9 16.9c7.6 7.6 20.5 2.2 20.5-8.5V76c0-6.6-5.4-12-12-12zM144 384c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
            />
          </svg>
          Gato Teste
        </h2>
        <div
          class="mt-2 space-y-1 text-sm font-medium text-gray-600"
          data-testid="cat-details"
        />
      </div>
    </a>
  </article>
</div>
`;
